doc/chips/9.rst

```diff
@@ -45,36 +49,285 @@ The idea of ``mason`` is based heavily off of Rust's cargo_. It is meant to
+    VisualDebug/
+        2.2.0.toml
+        2.2.1.toml
+
+Each versioned manifest file would be identical to the manifest file in the
+top-level directory of the package repository, with one additional field that
+is not required in the repository manifest, a URL pointing to the repository
+and revision in which the version is located.
+
+Continuing the example from before, the 'registry' ``Mason.toml`` would include
+the additional ``source`` field:
+
+
+.. code-block:: ini
+
+   [package]
+   name = "hello_world"
+   version = "0.1.0"
+   authors = ["Brad Chamberlain <brad@chamberlain.com>"]
+   source = { git = "https://github.com/bradcray/hello_world", tag = "0.1.0" }
+
+   [dependencies]
+   Curl = "1.0.0"
+
+
+There are some recognized downsides of pursuing a GitHub repository as a registry:
+
+* It is not be feasible to cache packages centrally, compromising the
+  reliability of the entire package ecosystem.
+
+  * The packages themselves are distributed despite there being a central
+    registry of manifests.
+  * We can *not* guarantee that a given package version will always remain
+    unchanged or even always exist.
+
+    * Although some form of nightly/weekly testing could help mitigate this
+      issue.
+
+* It is not easy to allow users to submit packages without human intervention
+  (gate keeper accepting pull requests).
+
+* GitHub's native interface would severely inhibit discoverability of packages.
+
+  * This can circumvented by building a separate website for exploring the
+    package ecosystem.
+
+    * Julia does a great job at this in their Ecosystem-Pulse_.
+
+For these reasons, we would like to eventually pursue a truly central registry
+with cached packaged and a first-class interface.
+
+**Package Integrity**
+
+There will be two primary measures taken to ensure that submitted packages are
+valid and not of poor quality:
+
+* Continuous Integration (CI) tools such as Travis CI could run a suite of
+  basic functionality tests to ensure a submitted package is properly
+  setup. Tests might include:
+
+  * Downloading the package.
+  * Building the package.
+  * Running any tests that the package provides.
+
+* A set of core and trusted non-core Chapel community developers will serve
+  as gate keepers who will allow packages into the repository by merging
+  pull requests that pass all tests.
+
+  * Other non-automated review procedures can be done by the gate keepers,
+    but it would be ideal to keep this effort minimal if required at all.
+
+  * This is similar to Homebrew's approach to package submission.
+
+
+Manifest File -> Lock File
+++++++++++++++++++++++++++
+
+This will be a fully automated step in which mason does the following:
+
+* Parses the manifest file
+* Builds a dependency directed acyclic graph (DAG)
+* Performs any incompatible version resolution strategy
+* Serializes the DAG and outputs it into ``Mason.lock``
+
+The user should never have to edit a ``Mason.lock`` file manually.
```

Comment:
We could make this statement stronger with "should never edit", since it'd get out of sync with the manifest.